atc/db/resource.go

```diff
@@ -949,3 +1000,252 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+	case CausalityUpstream:
+		return upStreamCausalityQuery
+	default:
+		return ""
+	}
+}
+
+type resourceVersionKey struct {
+	resourceID int
+	versionID  int
+}
+
+// causalityBuilds figures out all the builds that are related to a particular resource version
+// This can include builds that were used the resource version (and its descendents) as an input,
+// and builds that generated some ancestor of the build that generated the resource version itself.
+func causalityBuilds(tx Tx, resourceID int, versionMD5 string, direction CausalityDirection) (map[int]*atc.CausalityBuild, map[int]*atc.CausalityJob, error) {
+	query := causalityQuery(direction)
+	// construct the job and build nodes. These are placed into a map for easy access down the line
+	rows, err := psql.Select("b.id", "b.name", "b.status", "j.id", "j.name").
+		Prefix(query, resourceID, versionMD5).
+		From("build_ids bi").
+		Join("builds b ON b.id = bi.build_id").
+		Join("jobs j ON b.job_id = j.id").
+		RunWith(tx).
+		Query()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	builds := make(map[int]*atc.CausalityBuild, 0)
+	jobs := make(map[int]*atc.CausalityJob, 0)
+
+	for rows.Next() {
+		var buildID, jobID int
+		var buildName, jobName, status string
+
+		err = rows.Scan(&buildID, &buildName, &status, &jobID, &jobName)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		job, found := jobs[jobID]
+		if !found {
+			job = &atc.CausalityJob{
+				ID:   jobID,
+				Name: jobName,
+			}
+		}
+		// since we only iterate over each build once, this should never insert any duplicates
+		job.BuildIDs = append(job.BuildIDs, buildID)
+
+		build := &atc.CausalityBuild{
+			ID:     buildID,
+			Name:   buildName,
+			JobId:  jobID,
+			Status: atc.BuildStatus(status),
+		}
+
+		builds[buildID] = build
+		jobs[jobID] = job
+	}
+
+	return builds, jobs, nil
+}
+
+func causalityResourceVersions(tx Tx, resourceID int, versionMD5 string, direction CausalityDirection, result *atc.Causality) error {
+	// Bitmap to filter out resources that are not part of a output-input chain. This is done to filter out any other "root" resources.
+	// For now, the causality view is only intersted in the "downstream" resources, not neccessarily "parallel input" resources
+	// (reverse is true for upstream causality). If this ever changes in the future, it would be trivial to remove this filtering and get
+	// a more "pipeline-like" causality view
+	resourceHasParent := make(map[int]bool)
+	resourceHasParent[resourceID] = true // the root resource is the only one that is not required to have a parent
+
+	buildAsChild := func(rv *atc.CausalityResourceVersion, build *atc.CausalityBuild) {
+		if !contains(build.ResourceVersionIDs, rv.ID) { // in case a build outputs the same resource multiple times
+			build.ResourceVersionIDs = append(build.ResourceVersionIDs, rv.ID)
+		}
+
+		if used, found := resourceHasParent[rv.ResourceID]; !used || !found {
+			resourceHasParent[rv.ResourceID] = false
+		}
```

Comment:
Pretty sure this can just be removed - given that if a key isn't found in a `map`, the zero value is returned, the condition should be equivalent to:

```go
if !resourceHasParent[rv.ResourceID] {
	resourceHasParent[rv.ResourceID] = false
}
```

...which is redundant. We can just not store resources in the map if they don't have a parent for the same result.